### PR TITLE
add {pre,post}_visit_query to Visitor

### PR DESCRIPTION
**Pull request by jmhain**
Title says it all.

### CODE REVIEW

File: derive/src/lib.rs [hunk 1]
```diff
@@ -111,19 +111,48 @@ impl Attributes {
     }
 
     /// Returns the pre and post visit token streams
-    fn visit(&self, s: TokenStream) -> (Option<TokenStream>, Option<TokenStream>) {
+    fn visit(&self, s: TokenStream, is_option: bool) -> (Option<TokenStream>, Option<TokenStream>) {
```
Comment (alamb) [inline derive/src/lib.rs:114]: Can you please also document what `is_option` means in this contex?

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @jmhain  -- can you please add a test for this new functionality?

**Comment by jmhain** [comment]
@alamb Sorry for not including the tests in the original PR, just wanted to check that this funcitonal

**Comment by jmhain** [comment]
@alamb Sorry for not including tests in the original PR, I just wanted to make sure this functionality was desired first. :) In adding the tests I found a few bugs that resulted from me misunderstanding how the visit macro works. If you include the `#[visit]` attribute on both the type and the field, then the visitor methods actually end up being called twice. My recent PR that added `UNPIVOT` support had the same issue, as did the existing `PIVOT` table factor that I had used as a reference.

I've fixed each of these, added tests that previously reproduced the redundant calls, and added some additional details to the `README.md` for the derive crate to ensure that others don't make the same error.

File: src/ast/query.rs [hunk 2]
```diff
@@ -739,7 +740,6 @@ pub enum TableFactor {
     /// For example `FROM monthly_sales PIVOT(sum(amount) FOR MONTH IN ('JAN', 'FEB'))`
     /// See <https://docs.snowflake.com/en/sql-reference/constructs/pivot>
     Pivot {
-        #[cfg_attr(feature = "visitor", visit(with = "visit_table_factor"))]
```
Comment (alamb) [inline src/ast/query.rs:742]: why were these removed?

**Review by alamb — COMMENTED** [review]
Thank you @jmhain  -- this looks like a nice contribution to me. @tustvold  do you have a few minutes to review this?

File: derive/src/lib.rs (hunk 1 above)
Comment (jmhain) [reply derive/src/lib.rs]: The changes to the derive macro turned out not to be necessary

File: src/ast/query.rs (hunk 2 above)
Comment (jmhain) [reply src/ast/query.rs:742]: `TableFactor` is already annotated with this, so it results in two calls to both the `pre_` and `post_visit_table_factor` for the same table factor, e.g. this is the output of one of the new tests I added without this change:

```
 [
     "PRE: STATEMENT: SELECT * FROM monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d) ORDER BY EMPID",
     "PRE: QUERY: SELECT * FROM monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d) ORDER BY EMPID",
     "PRE: TABLE FACTOR: monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d)",
     "PRE: TABLE FACTOR: monthly_sales",
     "PRE: TABLE FACTOR: monthly_sales",
     "PRE: RELATION: monthly_sales",
     "POST: RELATION: monthly_sales",
     "POST: TABLE FACTOR: monthly_sales",
     "POST: TABLE FACTOR: monthly_sales",
     "PRE: EXPR: SUM(a.amount)",
     "PRE: EXPR: a.amount",
     "POST: EXPR: a.amount",
     "POST: EXPR: SUM(a.amount)",
     "POST: TABLE FACTOR: monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d)",
     "PRE: EXPR: EMPID",
     "POST: EXPR: EMPID",
     "POST: QUERY: SELECT * FROM monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d) ORDER BY EMPID",
     "POST: STATEMENT: SELECT * FROM monthly_sales PIVOT(SUM(a.amount) FOR a.MONTH IN ('JAN', 'FEB', 'MAR', 'APR')) AS p (c, d) ORDER BY EMPID",
 ]
```

Note the duplicate calls for `TABLE FACTOR: monthly_sales"`. 

I added a section in the `README.md` for the derive crate that explains this in more detail: https://github.com/jmhain/sqlparser-rs/blob/visit_query/derive/README.md

I can split this into a separate PR if you'd like since it's separate from adding `{pre,post}_visit_query`, I just discovered it now because I almost introduced another instance of this same bug here.

**Review by alamb — APPROVED** [review]
Thank you very much @jmhain
